nifi-nar-bundles/nifi-standard-services/nifi-dbcp-service-bundle/nifi-dbcp-service/src/test/java/org/apache/nifi/dbcp/HikariCPConnectionPoolTest.java

```diff
@@ -0,0 +1,273 @@
+        final Connection connection = dbcpService.getConnection();
+        Assert.assertNotNull(connection);
+
+        createInsertSelectDrop(connection);
+
+        connection.close(); // return to pool
+    }
+
+    @Rule
+    public ExpectedException exception = ExpectedException.none();
+
+    /**
+     * Test get database connection using Derby. Get many times, after a while pool should not contain any available connection and getConnection should fail.
+     */
+    @Test
+    public void testExhaustPool() throws InitializationException {
+        final TestRunner runner = TestRunners.newTestRunner(TestProcessor.class);
+        final HikariCPConnectionPool service = new HikariCPConnectionPool();
+        runner.addControllerService("test-exhaust", service);
+
+        // set embedded Derby database connection url
+        runner.setProperty(service, HikariCPConnectionPool.DATABASE_URL, "jdbc:derby:" + DB_LOCATION + ";create=true");
+        runner.setProperty(service, HikariCPConnectionPool.DB_USER, "tester");
+        runner.setProperty(service, HikariCPConnectionPool.DB_DRIVERNAME, "org.apache.derby.jdbc.EmbeddedDriver");
+
+        runner.enableControllerService(service);
+
+        runner.assertValid(service);
+        final DBCPService dbcpService = (DBCPService) runner.getProcessContext().getControllerServiceLookup().getControllerService("test-exhaust");
+        Assert.assertNotNull(dbcpService);
+
+        exception.expect(ProcessException.class);
+        exception.expectMessage("Connection is not available");
+        for (int i = 0; i < 100; i++) {
+            final Connection connection = dbcpService.getConnection();
+            Assert.assertNotNull(connection);
+        }
+    }
+
+    /**
+     * Test get database connection using Derby. Get many times, release immediately and getConnection should not fail.
+     */
+    @Test
+    public void testGetManyNormal() throws InitializationException, SQLException {
+        final TestRunner runner = TestRunners.newTestRunner(TestProcessor.class);
+        final HikariCPConnectionPool service = new HikariCPConnectionPool();
+        runner.addControllerService("test-exhaust", service);
+
+        // set embedded Derby database connection url
+        runner.setProperty(service, HikariCPConnectionPool.DATABASE_URL, "jdbc:derby:" + DB_LOCATION + ";create=true");
+        runner.setProperty(service, HikariCPConnectionPool.DB_USER, "tester");
+        runner.setProperty(service, HikariCPConnectionPool.DB_PASSWORD, "testerp");
+        runner.setProperty(service, HikariCPConnectionPool.DB_DRIVERNAME, "org.apache.derby.jdbc.EmbeddedDriver");
+
+        runner.enableControllerService(service);
+
+        runner.assertValid(service);
+        final DBCPService dbcpService = (DBCPService) runner.getProcessContext().getControllerServiceLookup().getControllerService("test-exhaust");
+        Assert.assertNotNull(dbcpService);
+
+        for (int i = 0; i < 1000; i++) {
```

Comment:
Testing `1000` iterations seems very high for a test method, can this be reduced to something much smaller?

Reply:
Took a zero off it :)